File: code/regex/perl/regex.pl
#! /usr/bin/perl
#
# Nachweis der potentiell exponentiellen Laufzeit des Regex Matchers von Perl
#
# Der regulaere Ausdruck
#
#      a?a?a?a?a?a?aaaa
#
# beziehungsweise seine Verallgemeinerung (a?)^na^n akzeptiert den
# String a^n, aber ein NEA wird dafuer Zeit O(2^n) benoetigen.
# 
# Das Programm erzeugt nacheinander regulaere Ausdruecke und passende
# Strings immer groesserer Laenge. Dabei kann man das exponentielle
# Anwachsen der Laufzeit beobachten.
#
# Ausfuehrung:
#
#  $ perl regex.pl
#
# (c) 2010 Prof Dr Andreas Mueller, Hochschule Rapperswil
#
use Time::HiRes qw(time);
$string = "";
$regex = "";

# Test den Regex Matcher fuer Stringlaengen zwischen 1 und 29
for ($i = 1; $i <= 29; $i++) {
	# Verlaengere die Komponenten $regex und $string um je ein
	# zusaetzliches Element
	$string .= "a";
	$regex .= "a?";

	# Baue daraus jetzt den regulaeren Ausdruck auf
	$r = $regex.$string;

	# Fuehre den Regex Match durch
	$start = time;
	if ($string =~ m/$r/) {
		$end = time;
		printf("%29s matches %s %*s %.6f\n", $string, $r, 3*(29-$i), "", $end - $start);
	} else {
		printf("%29s does not match %s\n", $string, $r);
	}
}

exit 0;

File: code/regex/go/regex.go
package main

import (
  "fmt"
  "regexp"
  "time"
)

func main() {
  r1 := ""
  r2 := ""

  for i := 0; i < 29; i++ {
    r1 += "a?"
    r2 += "a"
    regex := r1 + r2
    r, err := regexp.Compile(regex)

    if err != nil {
      fmt.Println(err)
    }

    start := time.Now()
    if r.MatchString(r2) {
      fmt.Printf("%8d us : %s matches %s\n", time.Since(start).Microseconds(), r2, regex)
    }
  }
}
